Compiler prints a "type" in a diagnostic that isn't a valid type

I am trying to use gtk from zig code, cimporting gtk.h. A number of gtk routines take variable numbers of arguments, which, if I try to call directly gets me

```
Semantic Analysis [604/983] ./book.zig:45:24: error: TODO: support C ABI for more targets. https://github.com/ziglang/zig/
                       -1);
```
To try to get around this and other problems, I have a set of interface routines corresponding to the gtk routines that cause trouble, written in C. I call those from zig and they call the corresponding gtk routine. Fine.

One of those routines needs to be passed a uuid, which is stored in a variable, call it foo, of type `[:0]const u8`. The parameter in the C routine is `char *`, of course. Clearly passing the slice as an argument was not going to work, so I tried `&foo[0]`. I first tried `[*:0]const u8` as the type of the parameter in the extern for the C routine. That resulted in
```
zig build-exe --main-pkg-path .. --single-threaded --object utilities.o `pkg-config --libs gtk+-3.0` -L/usr/local/lib -L/u
Semantic Analysis [603/976] ./book.zig:46:93: error: expected type '[*:0]const u8', found '*:0const u8'
                       constants.account_tree_store_guid_index, &foo[0],
```
I thought `*:0const u8` was odd-looking and confusing. I am already wrestling with understanding
the concept of zig pointers from the documentation, which I find completely inadequate (when I
have time, I will start to post issues explaining my confusion  and to the extent I can untangle it, suggestions for improving the documentation). I could not relate this to anything in the pointer documentation. But trying to get this code to compile, I changed the type of the parameter in the extern for the interface function to the suggested type, `*:0const u8`. That got me this:

```
zig build-exe --main-pkg-path .. --single-threaded --object utilities.o `pkg-config --libs gtk+-3.0` -L/usr/local/lib -L/usr/lib -lc -lsqlite3 newcash.zig
./book.zig:13:63: error: invalid character: 'c'
    name_index:u8, name:[*:0]const u8, guid_index:u8, guid:*:0const u8, flags_index:u8, flags:u32) void;
```
So the right hand won't accept something suggested by the left-hand.

This is my second attempt at doing this project in zig. I abandoned it the first time because it was clear that zig simply wasn't mature enough, however great I thought its potential was. Many months later, I'm trying again and have made a lot more progress. I am still running into some issues, but the primary problem is the pointer documentation. It is a key part of the language and the unifying concept(s) behind it are simply not explained, so it's difficult, at least for me, to have a mental model of what should be written in specific situation and what the compiler expects. So everything becomes a special case. I'm going to put this work on hold again because it's just too much fighting with the language rather than getting my work done. I will check in again after a suitable hiatus.

--- a/src/analyze.cpp
+++ b/src/analyze.cpp
@@ -578,6 +578,7 @@ ZigType *get_pointer_to_type_extra2(CodeGen *g, ZigType *child_type, bool is_con
     }
     switch (ptr_len) {
         case PtrLenSingle:
+            assert(sentinel == nullptr);
             buf_appendf(&entry->name, "*");
             break;
         case PtrLenUnknown:

--- a/src/ir.cpp
+++ b/src/ir.cpp
@@ -21078,7 +21078,7 @@ static ZigType *adjust_ptr_len(CodeGen *g, ZigType *ptr_type, PtrLen ptr_len) {
             ptr_type->data.pointer.allow_zero,
             ptr_type->data.pointer.vector_index,
             ptr_type->data.pointer.inferred_struct_field,
-            ptr_type->data.pointer.sentinel);
+            (ptr_len != PtrLenUnknown) ? nullptr : ptr_type->data.pointer.sentinel);
 }
 
 static ZigType *adjust_ptr_allow_zero(CodeGen *g, ZigType *ptr_type, bool allow_zero) {

--- a/test/stage1/behavior/pointers.zig
+++ b/test/stage1/behavior/pointers.zig
@@ -1,6 +1,7 @@
 const std = @import("std");
-const expect = std.testing.expect;
-const expectError = std.testing.expectError;
+const testing = std.testing;
+const expect = testing.expect;
+const expectError = testing.expectError;
 
 test "dereference pointer" {
     comptime testDerefPtr();
@@ -330,4 +331,9 @@ test "@ptrToInt on null optional at comptime" {
         const pointer = @intToPtr(?*u8, 0xf00);
         comptime expect(0xf00 == @ptrToInt(pointer));
     }
+}
+
+test "indexing array with sentinel returns correct type" {
+    var s: [:0]const u8 = "abc";
+    testing.expectEqualSlices(u8, "*const u8", @typeName(@TypeOf(&s[0])));
 }

